test-harness/src/main/java/io/jenkins/plugins/casc/misc/junit/jupiter/AbstractRoundTripTest.java

```diff
@@ -0,0 +1,215 @@
+
+    /**
+     * Return the resource path (yaml file) to be loaded. i.e: If the resource is in the same
+     * package of the implementor class, then: my-config.yaml
+     *
+     * @return the resource name and path.
+     */
+    protected String configResource() {
+        return "configuration-as-code.yaml";
+    }
+
+    /**
+     * Return the string that should be in the logs of the JCasC logger to verify it's configured
+     * after a restart. This string should be unique to avoid interpreting that it was configured
+     * successfully, but it wasn't.
+     *
+     * @return the unique string to be in the logs to certify the configuration was done
+     * successfully.
+     */
+    protected abstract String stringInLogExpected();
+
+    /**
+     * 1. Configure the instance with the {@link #configResource()} implemented. 2. Check it was
+     * configured correctly. 3. Check the configuration is valid via Web UI. 4. Apply the
+     * configuration via Web UI. 5. Write the configuration to $JENKINS_HOME/jenkins.yaml. 6.
+     * Restart Jenkins. 7. Check the {@link #stringInLogExpected()} is set during the restart.
+     *
+     * @throws IOException If an exception is thrown managing resources or files.
+     */
+    @Test
+    public void roundTripTest(JenkinsRule r) throws Throwable {
+        String resourcePath = configResource();
+        String resourceContent = getResourceContent(resourcePath);
+
+        assertNotNull(resourcePath);
+        assertNotNull(resourceContent);
+
+        // Configure and validate
+        configureWithResource(resourcePath);
+        assertConfiguredAsExpected(r, resourceContent);
+
+        // Check config is valid via Web UI
+        String jenkinsConf = getResourceContent(resourcePath);
+        assertConfigViaWebUI(r, jenkinsConf);
+
+        // Apply configuration via Web UI
+        applyConfigViaWebUI(r, jenkinsConf);
+        assertConfiguredAsExpected(r, resourceContent);
+
+        // Configure Jenkins default JCasC file with the config file. It's already established, we check if applied
+        // looking at the logs.
+        putConfigInHome(r, jenkinsConf);
+
+        // Start recording the logs just before restarting, to avoid capture the previous startup. We're look there
+        // if the "magic token" is there
+        List<String> messages = new ArrayList<>();
+        Logger logger = Logger.getLogger("io.jenkins.plugins.casc");
+        logger.addHandler(new ConsoleHandler() {
+            @Override
+            public void publish(LogRecord record) {
+                super.publish(record);
+
+                String message = new SimpleFormatter().formatMessage(record);
+                Throwable x = record.getThrown();
+                synchronized (messages) {
+                    messages.add(message == null && x != null ? x.toString() : message);
+                }
+            }
+        });
+        logger.setLevel(Level.ALL);
```

Comment:
This is a replacement for using `org.jvnet.hudson.test.LoggerRule`. I will look into providing a JUnit5-compatible replacement for that as well. In the meantime this does more or less the same.

Reply:
Replaced by `LogRecorder`